### ISSUE TITLE
Injector class not serializable

### ISSUE THREAD

**Issue by rahul22048** [Urigo/graphql-modules#754]
I was trying to serialize injector class in my resolver in order to execute it in a separate process or a thread. But in trying to do so i got the following error:

```
Converting circular structure to JSON
  --> starting at object with constructor 'Socket'
    |     property 'parser' -> object with constructor 'HTTPParser'
        --- property 'socket' closes the circle
```

This usually happens when there is a circular dependency inside the class. Can this be fixed?

**Comment by ardatan** [comment]
@rahul22048 `Injector` has its own methods inside prototype so it is not possible to execute those methods in another process.

**Comment by rahul22048** [comment]
There are many open source libraries which serialize objects who have functions attached to their prototypes. I have also tested them with classes having functions. They fail in case the instance you are trying to serialize has some sort of cyclic dependency.

**Comment by ardatan** [comment]
Could you give me a working example how to use the functioms inside serialized object so we can implement this in GraphQL Modules? Also PRs are welcome :)

**Comment by rahul22048** [comment]
inside my resolver:-

userBehance: async (root, args, {injector}) => {
  let obj = serialize(injector);// error coming in this line only. Error text is same as above.
// after this i spin up a new thread and pass serialized obj inside it. Following code runs inside a //separate thread.
}
  let deserializedObj = deserialize(obj);
  let provider = deserializedObj.get(TestProvider);
// then i can proceed ahead with my logic ahead..

**Comment by rahul22048** [comment]
On a side note i also tried with context obj, there also i got the same error.

**Comment by ardatan** [comment]
Right now, we don't support something like this. And I don't think it is possible to call a function in another process using JSON's parse and stringify. If you have a working example without GraphQL Modules, please share so maybe we can find a way to make it run with GraphQL Modules.

**Comment by rahul22048** [comment]
Yes you are correct it is not possible to do it with JSON parse and stringify. The serialize function i have used above is of some open source library which works fine for ES 6 classes with functions inside class body and are without cyclic dependency. You can check it in your injector class may be something to do with the order of imports or may be external dependencies are having some form of cyclic dependency.

**Comment by ardatan** [comment]
@rahul22048 Could you share it?

**Comment by rahul22048** [comment]
I can't share my actual production code, but i can share with you the library that i used. PFB the github link for their repo:- https://github.com/yahoo/serialize-javascript.

This works for normal ES 6 classes.

Also note that i also tried with few other libraries as well, but all were giving error.

**Comment by rahul22048** [comment]
@ardatan  do you have any update on the issue?

**Comment by ardatan** [comment]
I checked that library. I don't think it is possible to use that library with our Injector because we have WeakMaps, references to the network object, class references and other stuff in `Injector` class. And I am not sure if this library supports prototypes and other stuff. I recommend you to use `cluster` feature of NodeJS instead to have multi-processes which look more relevant solution for your case;
https://nodejs.org/api/cluster.html

**Comment by rahul22048** [comment]
Clustering spins up an entire application on a separate process, whereas i want isolation at the module level.I want to spin a separate process for a module not for an entire application. Secondly i understand your class has some external dependencies, but what is causing the circular dependency error?  Can you find where is circular dependency inside that injector class?

**Comment by kamilkisiela** [comment]
We released a new major version (v1.0) of GraphQL Modules. Here's the [migration guide](https://graphql-modules.com/docs/next/recipes/migration).

I'm closing the thread.

Please try to upgrade. If you can't and still want to use the legacy version, please open a new issue.
